### PR TITLE
[ColorUtil] New methods kelvinToXY and xyToKelvin

### LINKED ISSUE #4369: [ColorUtil] Add conversions between colour temperature and CIE XY colour space coordinates
There has been a discussion on the forum about a need in the Zigbee binding for converting a colour temperature in Kelvin to the respective coordinate on the colour temperature locus in the CIE colour space, and vice-versa.

See https://community.openhab.org/t/dimmer-and-color-item-to-same-channel/157241

Signed-off-by: AndrewFG [software@whitebear.ch](mailto:software@whitebear.ch)

### PR DESCRIPTION
**Pull request by andrewfg**
Resolves #4369

This PR provides two new methods in the `ColorUtils` class as follows:
1. New method `kelvinToXY()` to convert a colour temperature in Kelvin to a point on the colour temperature locus in the CIE XY colour space, as described [here](https://www.waveformlighting.com/tech/calculate-cie-1931-xy-coordinates-from-cct)
2. New method `xyToKelvin()` to convert a point on the colour temperature locus in the CIE XY colour space to a colour temperature in Kelvin, as described [here](https://www.waveformlighting.com/tech/calculate-color-temperature-cct-from-cie-1931-xy-coordinates)

image

Note: the PR is created in ColorUtils in OH Core in case these conversions might be needed in other bindings.

Signed-off-by: AndrewFG <software@whitebear.ch>

### CODE REVIEW

**Comment by holgerfriedrich** [comment]
Maybe some more background information, in case the sources given above are no longer available....
https://en.wikipedia.org/wiki/Correlated_color_temperature#Approximation

**Comment by andrewfg** [comment]
See https://github.com/openhab/org.openhab.binding.zigbee/pull/855

**Comment by holgerfriedrich** [comment]
@andrewfg is this still a draft or shall I have a look?

File: bundles/org.openhab.core/src/main/java/org/openhab/core/util/ColorUtil.java [hunk 1]
```diff
@@ -650,4 +650,530 @@ public Point closest(Point p) {
...
+
+    /**
+     * Search the 'KELVIN_TO_XY_LOOKUP_TABLE' for the XY entry closest to the given colour temperature.
+     * Uses a recursive 'QuickSearch' algorithm.
+     * 
+     * @param kelvin the colour temperature in K to find
+     * @param min the first index in the lookup table
+     * @param max the last index in the lookup table
+     * @return an array with the found CIE colour XY values
+     * @throws IndexOutOfBoundsException
```
Comment (holgerfriedrich) [inline bundles/org.openhab.core/src/main/java/org/openhab/core/util/ColorUtil.java:1126]: description is missing, causes a warning

**Review by holgerfriedrich — COMMENTED** [review]
Overall: LGTM.
Personally, I do not like that huge lookup table in the code, but I see the function is not easily invertible. So probably the approach is fine.

**Comment by andrewfg** [comment]
> is this still a draft or shall I have a look?

I was keeping it as draft while writing the respective (first) addon PR .. in case there may be learnings from it. ..

**Comment by andrewfg** [comment]
> do not like that huge lookup table

Honestly me neither. But as you say the function is difficult to invert.

File: bundles/org.openhab.core/src/main/java/org/openhab/core/util/ColorUtil.java (hunk 1 above)
Comment (andrewfg) [reply bundles/org.openhab.core/src/main/java/org/openhab/core/util/ColorUtil.java]: fixed

**Review by holgerfriedrich — APPROVED** [review]

**Comment by andrewfg** [comment]
Many thanks @holgerfriedrich :)

**Comment by holgerfriedrich** [comment]
Do you think we should expose this to DSL rules (core repo org.openhab.core.model.script/src/org/openhab/core/model/script/actions/CoreUtil.java; docs repo, addons/actions.md) and to javascript?

**Comment by andrewfg** [comment]
@holgerfriedrich are any of the other public static methods of the ColorUtils class exposed? If so then I suggest to be simply consistent across all such methods.

**Pull request merged**